test/test_torch.py

```diff
@@ -6987,80 +6988,74 @@ def test_logical_and(self, device):
-        torch.clamp(m1, min=min_val, max=max_val, out=out)
-        self.assertEqual(out, res1)
+        if isinstance(max_vals, torch.Tensor):
+            max_vals = max_vals.cpu().numpy()
 
-        res1 = torch.clamp(m1, min=min_val)
-        res2 = m1.clone()
-        for i in iter_indices(res2):
-            res2[i] = max(min_val, res2[i])
-        self.assertEqual(res1, res2)
+        # Use NumPy implementation as reference
+        X_clamped = torch.tensor(np.clip(X.cpu().numpy(), a_min=min_vals, a_max=max_vals))
+        return X, X_clamped
 
-        torch.clamp(m1, min=min_val, out=out)
-        self.assertEqual(out, res1)
+    def test_clamp(self, device):
+        # Create min/max argument product
+        args = product((torch.linspace(-1, 1, 100, device=device), -1, None),
+                       (torch.linspace(-1, 1, 100, device=device), 1, None))
 
-        res1 = torch.clamp(m1, max=max_val)
-        res2 = m1.clone()
-        for i in iter_indices(res2):
-            res2[i] = min(max_val, res2[i])
-        self.assertEqual(res1, res2)
+        for min_val, max_val in args:
+            if min_val is None and max_val is None:
+                continue
 
-        torch.clamp(m1, max=max_val, out=out)
-        self.assertEqual(out, res1)
+            X, Y_expected = self.generate_clamp_baseline(device, min_val, max_val)
 
-        # if the tensor contains nan case
-        test_tens = torch.tensor([nan], device=device)
+            Y_actual = torch.clamp(X, min_val, max_val)
+            self.assertEqual(Y_expected, Y_actual)
 
-        res1 = test_tens.clone()
-        res1.clamp_(min_val, max_val)
-        res2 = test_tens.clone()
-        for i in iter_indices(res2):
-            res2[i] = max(min(res2[i], max_val), min_val)
-        self.assertEqual(torch.isnan(res1), torch.isnan(res2))
+            Y_out = torch.empty_like(X)
+            torch.clamp(X, min=min_val, max=max_val, out=Y_out)
+            self.assertEqual(Y_expected, Y_out)
 
-        out = test_tens.clone()
-        torch.clamp(test_tens, min=min_val, max=max_val, out=out)
-        self.assertEqual(torch.isnan(out), torch.isnan(res1))
+            X.clamp_(min_val, max_val)
+            self.assertEqual(Y_expected, X)
 
-        res1 = torch.clamp(test_tens, min=min_val)
-        res2 = test_tens.clone()
-        for i in iter_indices(res2):
-            res2[i] = max(res2[i], min_val)
-        self.assertEqual(torch.isnan(res1), torch.isnan(res2))
+    def test_clamp_raises_arg_errors(self, device):
+        X = torch.randn(100, dtype=torch.float, device=device)
+        error_msg = 'At least one of \'min\' or \'max\' must not be None'
+        with self.assertRaisesRegex(RuntimeError, error_msg):
+            X.clamp()
+        with self.assertRaisesRegex(RuntimeError, error_msg):
+            X.clamp_()
+        with self.assertRaisesRegex(RuntimeError, error_msg):
+            torch.clamp(X)
+
+    def test_clamp_propagates_nans(self, device):
```

Comment:
I appreciate how thorough this is but do we actually need this test? What if `test_clamp` sets `with_nan=True`? (Unless NumPy doesn't propagate NaNs).

Reply:
I think it is necessary and it is not new - it was there before as well. The reason for test this is because different hardware/backends handle nan propagation differently. In particular, I had to make sure that ROCm backend propagated nans from the main input tensor. 

NumPy propagates nans (and this test uses NumPy as reference) but I think it is best to keep it separate from the other test because it tests a different part of functionality: `test_clamp` makes sure that the outputs are between min and max, and `test_clamp_propagates_nans` tests whether nan values in the input tensor are _always_propagated and are not masked by min/max. 

Moreover, keeping them separate allows at certain times disabling the nan test and still testing the main functionality. It has been, for example, disabled in XLA, since there is a known XLA behaviour difference in nan propagation (see above in the PR, where it was discussed with XLA team members), that is non-trivial to fix.

Reply:
OK. Sounds reasonable.